### PR TITLE
Paper: Yaksh: Facilitating Learning by Doing

### PR DESCRIPTION
**Pull request by prabhuramachandran**
A first draft of the paper "Yaksh: facilitating learning by doing".

### CODE REVIEW

**Comment by prabhuramachandran** [comment]
@deniederhut -- I see that today is the last day for the preliminary review but don't see any comments on this submission.  Did we miss something?

**Comment by deniederhut** [comment]
Hey Prabhu! The deadline for reviews is this Thursday. We've been doing our
best to remind reviewers to take a look at their papers, and will make sure
that every submission gets at least one review by next week.

**Comment by prabhuramachandran** [comment]
OK thank you!

**Comment by msarahan** [comment]
This looks good to me.  One thing that was not clear to me was whether coding tests could be evaluated on multiple test conditions.  Several of the online test platforms I've encountered have test conditions that use different input to the code to test different edge cases (large data, algorithmic complexity, data pathology).  That seems much more rigorous than a single simple pass/fail on one input, and examples of how Yaksh supports or might support that in the paper would be helpful.

**Comment by prabhuramachandran** [comment]
@msarahan -- Thank you.  Yes indeed, you can have as many tests as you want.  In fact we also support partial grading, so a user gets partial marks if they only pass some tests.  We will modify the paper to clearly mention this.

**Comment by mfenner1** [comment]
Prabhu, my apologies for being late with my review.  I will have comments here today or tomorrow.  I have the manuscript in hand now and I'm looking forward to reading it.

**Comment by jwjohnson314** [comment]
This is an interesting project and one I'm interested in using for my own classes. The paper is well written and thorough. There is a typo on page 3 (column 1, '... one can create a full fleged...' should be '...one can create a full fledged...'). Other than that I think it looks great and am glad I had the opportunity to review it.

File: papers/yaksh/yaksh.rst [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+:institution: FOSSEE IIT Bombay, Mumbai, India
+
+
+--------------------------------------
+Yaksh: Facilitating Learning by Doing
+--------------------------------------
+
+.. class:: abstract
+
+    Yaksh is a free, and open-source online evaluation platform. At its core,
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: free, open

File: papers/yaksh/yaksh.rst [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+
+Introduction
+-------------
+
+Yaksh_ is created by the `FOSSEE Python team`_. The `FOSSEE project`
+(http://fossee.in) based at IIT Bombay, is funded by the Ministry of Human
+Resources and Development, MHRD (http://mhrd.gov.in) of the Government of
+India. The goal of the FOSSEE project is to increase the adoption of Free and
+Open Source Software in Education in India. The project started in 2009 and
+develops, and promotes a variety of open source projects. FOSSEE's Python
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: in 2009 to develop and promote ....

File: papers/yaksh/yaksh.rst [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+.. _`FOSSEE project`: https://fossee.in
+.. _Yaksh: https://github.com/FOSSEE/online_test
+
+
+Overview of Yaksh
+---------------------
+
+Since the emergence of learning management system (LMS) and massive open
+online course (MOOC) providers, e-learning has grown significantly. Despite
+the ever increasing adopters, major platforms still use simple question types
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: simple questions like

File: papers/yaksh/yaksh.rst [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
+like multiple-choice questions and uploading of assignments from students as a
+means to evaluate students' performance. Yaksh seeks to improve on this.
+
+It is well known that practice assignments and problem solving improve
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: I'd simplify this paragraph substantially.

File: papers/yaksh/yaksh.rst [hunk 5]
```diff
@@ -0,0 +1,937 @@
...
+In 2011, the first version of Yaksh was developed to administer programming
+quizzes for an online teacher training course that FOSSEE conducted. More than
+600 teachers were trained and we wanted them to be able to write programs and
+have those corrected. This work was presented in at SciPy India 2011 [PR11]_.
+It would have been impossible to do this manually.
+
+Yaksh is a free, and open-source online evaluation software that allows
+teachers to create courses and students to watch lessons and attempt tests
+which are evaluated immediately. Yaksh is designed to be used by a large
+number of users concurrently thereby making it apt for using in schools,
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: apt for use in

File: papers/yaksh/yaksh.rst [hunk 6]
```diff
@@ -0,0 +1,937 @@
...
+colleges and other educational institutes for training a large number of
+students.
+
+Yaksh is implemented in Python and uses Django
+(https://www.djangoproject.com/). It is also written as a pip-installable
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: It is deployed as a pip-installable / It can be deployed via pip ....

File: papers/yaksh/yaksh.rst [hunk 7]
```diff
@@ -0,0 +1,937 @@
...
+within their project. The sources are available from:
+https://github.com/FOSSEE/online_test
+
+To use Yaksh, one could sign-up on the official https://yaksh.fossee.in
+website or host it on one's own servers. The most standard and secure way to
+deploy Yaksh on a server is to build separate docker images using docker
+compose. Instructions for this are available in the yaksh sources and are easy
+to setup.
+
+For teachers, Yaksh provides a wide array of question types which include the
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: which include:  multiple choice, .....

File: papers/yaksh/yaksh.rst [hunk 8]
```diff
@@ -0,0 +1,937 @@
...
+basic question types like multiple choice, fill-in-the-blanks, assignment
+upload, etc. One can also add standard I/O and assertion-based questions for
+simple and basic programming questions. For complex programs, teachers can add
+a hook-based test case which would enable them to take the student answer and
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: which enable

File: papers/yaksh/yaksh.rst [hunk 9]
```diff
@@ -0,0 +1,937 @@
...
+part of what Yaksh does.
+
+nbgrader_ is a Jupyter Notebook plugin that can be used to grade programming
+assignments. The student submits jupyter notebooks containing code blocks
+which are then evaluated manually or automatically. nbgrader_ provides a very
+convenient Jupyter based interface. Yaksh offers very different features, it
+offers instant feedback and grading, supports a variety of different
+languages, and also allows one to host a full course.
+
+ .. _nbgrader: https://github.com/jupyter/nbgrader
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: This didn't display in my pdf version.

File: papers/yaksh/yaksh.rst [hunk 10]
```diff
@@ -0,0 +1,937 @@
...
+Python. It has been used in several courses taught at IIT Bombay and also for
+conducting recruitment interviews internally.
+
+There are a few other open source software packages that allow to do all or
+part of what Yaksh does.
+
+nbgrader_ is a Jupyter Notebook plugin that can be used to grade programming
+assignments. The student submits jupyter notebooks containing code blocks
+which are then evaluated manually or automatically. nbgrader_ provides a very
+convenient Jupyter based interface. Yaksh offers very different features, it
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: Yaksh offers instant feedback and grading ......

File: papers/yaksh/yaksh.rst [hunk 11]
```diff
@@ -0,0 +1,937 @@
...
+
+In this paper we first discuss how Yaksh may be installed, its features, and a
+high-level overview of its design and implementation. We then present some
+information on how Yaksh has been used at FOSSEE for a variety of tasks.
+
+
+Installation and setup
+----------------------
+
+Deployment of a web application for temporary use as well as for production
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: temporary use as well as for  --> evaluation and

File: papers/yaksh/yaksh.rst [hunk 12]
```diff
@@ -0,0 +1,937 @@
...
+However, this is not recommended since this leaves the base system potentially
+vulnerable to malicious code. In case one wishes to use this method, all
+Python dependencies will have to be installed using ``sudo``.
+
+In order to access the interface, one can run the web server using::
+
+  $ invoke serve
+
+This command will run the django application server on the **8000** port and
+can be accessed using a browser, also this port will be unavailable
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: ---> broweser.  [drop rest of sentence]

File: papers/yaksh/yaksh.rst [hunk 13]
```diff
@@ -0,0 +1,937 @@
...
+-------------------------
+
+Once Yaksh is installed and running, one can create a full fleged course with
+lessons, practice, and evaluation based quizzes. Yaksh supports following
+languages such as Python, Java, C, C++, and Scilab. It provides several
+question types such as Single Correct Choice, Multiple Correct Choice,
+Programming, Fill in the blanks, Arrange the options, Assignment upload. For
+simple and complex questions several test case types are provided such as
+standard input/output test case, Standard Assertion test case, Hook based test
+case, MCQ based test case, etc. The interface provides instant feedback for
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: define MCQ on first use

File: papers/yaksh/yaksh.rst [hunk 14]
```diff
@@ -0,0 +1,937 @@
...
+web applications, handles basic security issues, and provides a
+basic authentication system.
+
+Django makes it easy to store information in a database by providing an
+object-relational mapping (ORM). This allows users to define the database
+tables at a very high level without having to write raw SQL queries.
+
+Django provides a view controller to handle all the requests sent from the
+client side. A view then interacts with the database if any database data is
+required, collects all the data and sends the data back to the templates which
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: all the necessary data

File: papers/yaksh/yaksh.rst [hunk 15]
```diff
@@ -0,0 +1,937 @@
...
+
+  This model is used for storing question information such as name,
+  description etc. Once the questions are created they are added in the
+  question paper
+
+- TestCase
+
+  This model is used for storing question test cases.
+
+  Different test case models are available which subclasses the TestCase
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: which subclass

File: papers/yaksh/yaksh.rst [hunk 16]
```diff
@@ -0,0 +1,937 @@
...
+  This model is used for storing answerpaper for a particular course and quiz.
+
+- Answer
+
+  This model is used for storing the answer submitted by the user which are
+  added to answerpaper.
+
+- Lesson
+
+  A lesson can be any markdown text with/or an embedded video of a particular
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: with/or is awkward for me

File: papers/yaksh/yaksh.rst [hunk 17]
```diff
@@ -0,0 +1,937 @@
...
+Code Server is an important part of yaksh. The evaluation of any code is done
+through the code server. We have used the Tornado_ web framework to manage the
+asynchronous process generation. A ``settings.py`` file is provided which
+is used to specify various parameters necessary for the code server.
+
+.. _Tornado: https://www.tornadoweb.org
+
+This settings file contains information such as:
+
+- number of code server processes required to process the code.
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst:483]: I would remove periods at the ends of these list items.

File: papers/yaksh/yaksh.rst [hunk 18]
```diff
@@ -0,0 +1,937 @@
...
+- server pool port, a common port for accessing the Tornado web server.
+- server host name, a server host for accessing the Tornado web server.
+- a timeout to prevent infinite loops locking up a process.
+- dictionary of code evaluators based on the programming language.
+
+
+A Tornado HTTP server is started with the specified server host and server
+pool port from the settings. The server handles POST requests which are
+submitted with a unique id, along with a JSONized string with the programming
+language of the question, the user answer, files (if any), test case data i.e.
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: the next four or five lines are a bit awkward to read

File: papers/yaksh/yaksh.rst [hunk 19]
```diff
@@ -0,0 +1,937 @@
...
+
+  To make the code evaluation more secure all the code evaluation is done
+  inside a docker container. Docker containers can also be used to limit the
+  use of system resources such as CPU utilization, memory utilization etc.
+
+
+Workflow of yaksh
+-----------------
+
+- **Authentication system**
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: I would consider pulling authentication out of the workflow section (just move it up above the prior heading).  then workflow can be instructor workflow and student workflow.

File: papers/yaksh/yaksh.rst [hunk 20]
```diff
@@ -0,0 +1,937 @@
...
+  student to attempt it for a while and then show a solution.
+
+  .. figure:: yaksh_question.png
+     :alt:  Question interface
+
+     Question interface :label:`fig:yaksh-question`
+
+  A programming question can have test case types of standard assert,
+  standard I/O or a hook.
+  Fig. :ref:`fig:yaksh-testcase` shows sample test case of standard assert
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: Figure 4 shows a sample test case

File: papers/yaksh/yaksh.rst [hunk 21]
```diff
@@ -0,0 +1,937 @@
...
+  creates a quiz and an associated question paper. A quiz can have a passing
+  criterion. Quizzes have active durations and each question paper will have a
+  particular time within which it must be completed. For example one could
+  conduct a 15 minute quiz with a 30 minute activity window.
+  Questions are automatically graded. A user either gets the full marks or
+  zero if the tests fail. If a question is allowed to have partial grading
+  then during evaluation the user gets partial marks based on passing
+  test cases.
+
+  The moderator can then create lessons and modules. A lesson can have
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: This paragraph seemed wordy to me.  I'd rewrite more directly.

File: papers/yaksh/yaksh.rst [hunk 22]
```diff
@@ -0,0 +1,937 @@
...
+  description either as a markdown text or a video or both. After lesson
+  creation, the moderator can create modules. A module can have its own
+  description either as a markdown text or a video or both. All the lessons
+  and quizzes are added to the created module. The moderator can create any
+  number of modules, lessons and quizzes as desired.
+  These modules are added to a course.
+
+
+  Fig. :ref:`fig:yaksh-monitor` shows monitor for a quiz from one of the
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: Fig 5 shows a monitor .... one of the courses

File: papers/yaksh/yaksh.rst [hunk 23]
```diff
@@ -0,0 +1,937 @@
...
+  course running on yaksh. The instructors can also monitor students in real
+  time during a quiz there by gaining insight on how students are performing.
+  The moderator can also view complete course progress showing student
+  progress for overall course. i.e how many modules a student has completed,
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: for the overall course such as the number and percent of completed modules.

File: papers/yaksh/yaksh.rst [hunk 24]
```diff
@@ -0,0 +1,937 @@
...
+  Python course.
+
+  .. figure:: yaksh_view_lesson.png
+     :alt:  Lesson
+
+     The interface showing a video lesson
+     :label:`fig:yaksh-lesson`
+
+
+  Fig. :ref:`fig:yaksh-mcq` shows a mcq question from a quiz. A student can
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: MCQ

File: papers/yaksh/yaksh.rst [hunk 25]
```diff
@@ -0,0 +1,937 @@
...
+
+  .. figure:: yaksh_error.png
+     :alt:  Error
+
+     Error output after submitting the code answer.
+     :label:`fig:yaksh-error`
+
+  Fig. :ref:`fig:yaksh-code-stdio` shows an StdIO based question. Once the
+  answer is submitted we get the error output as shown in
+  fig :ref:`fig:yaksh-error-stdio`. Fig :ref:`fig:yaksh-error-stdio` shows
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: fig --> Fig

File: papers/yaksh/yaksh.rst [hunk 26]
```diff
@@ -0,0 +1,937 @@
...
+
+
+  .. figure:: yaksh_error_stdio.png
+     :alt:  Error Stdio
+
+     Error output for stdio question type.
+     :label:`fig:yaksh-error-stdio`
+
+  Students can take the practice exercises where each question in the exercise
+  is timed. Student has to solve the question within the specified time,
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: Students must solve the question ....

File: papers/yaksh/yaksh.rst [hunk 27]
```diff
@@ -0,0 +1,937 @@
...
+
+  .. figure:: yaksh_stdio_que.png
+     :alt:  Programming Stdio
+
+     The interface for a stdio question type on yaksh.
+     :label:`fig:yaksh-code-stdio`
+
+  Students can submit the answer multiple times, thereby improving
+  their answers. Suppose a student is not able to solve a question, that
+  question can be skipped and can be attempted later. All the submitted and
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: can be skipped an attempted later.

File: papers/yaksh/yaksh.rst [hunk 28]
```diff
@@ -0,0 +1,937 @@
...
+  .. figure:: yaksh_error.png
+     :alt:  Error
+
+     Error output after submitting the code answer.
+     :label:`fig:yaksh-error`
+
+  Fig. :ref:`fig:yaksh-code-stdio` shows an StdIO based question. Once the
+  answer is submitted we get the error output as shown in
+  fig :ref:`fig:yaksh-error-stdio`. Fig :ref:`fig:yaksh-error-stdio` shows
+  the user output and expected output separately indicating line by line
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: indicating the line by line

File: papers/yaksh/yaksh.rst [hunk 29]
```diff
@@ -0,0 +1,937 @@
...
+
+- The ``teardown`` method is used to delete all the files that are not
+  relevant once the execution is done.
+
+- All the code compilation task will be done by the ``compile_code`` method.
+  There is no need to add this method if there is no compilation procedure.
+
+- The execution of the code is performed in the ``check_code`` method.
+
+This method should return three values-
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: The ``check_code`` method must return ....

File: papers/yaksh/yaksh.rst [hunk 30]
```diff
@@ -0,0 +1,937 @@
...
+  at IIT Bombay. Yaksh is used to teach Python to some undergraduate students.
+  These courses have served as a test-bed for the software. Thus far, about
+  300 students from IIT Bombay have been taught using Yaksh.
+
+
+- **Usage for Python Workshops**
+
+
+  In early 2017, FOSSEE started conducting remote Python workshops in
+  technical colleges across India. These workshops consists of several
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: These workshops consist

File: papers/yaksh/yaksh.rst [hunk 31]
```diff
@@ -0,0 +1,937 @@
...
+  sessions spread through one or three days, depending on the type of the
+  course an institute chooses. A session typically begins with screening a
+  video tutorial inside the venue. The tutorials are followed by a demanding
+  set of exercises and quizzes, both of which are conducted on Yaksh. This is
+  followed by brief Q&A sessions with the remotely assisting Python
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: with the remote Python instructors

File: papers/yaksh/yaksh.rst [hunk 32]
```diff
@@ -0,0 +1,937 @@
...
+  1. As of mid 2018, around 13,000 active users are on Yaksh, with more
+     expected to join by the end of the year.
+
+  2. Rapidly growing user base with about 730, 4500 and 7500 user
+     registrations for year 2016, 2017 and mid-2018 respectively.
+
+  3. 100+ institutes have conducted the workshop with about 6000 students
+     participating and about 3600 students obtaining a certificate.
+
+  4. For the first three months of the Python self learning Course, an
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: course

File: papers/yaksh/yaksh.rst [hunk 33]
```diff
@@ -0,0 +1,937 @@
...
+     estimate of 3500 students enrolled with 1200 completing the course within
+     the time frame and 400 students obtaining a passing certificate.
+
+  5. An equal ratio of male to female participants with most users from the
+     age of 18-24 as seen in the figures. :ref:`fig:yaksh-gender-demographics`
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: as seen in Figures

File: papers/yaksh/yaksh.rst [hunk 34]
```diff
@@ -0,0 +1,937 @@
...
+     and :ref:`fig:yaksh-age-demographics`.
+
+  6. Average time spent on the website by an user is around 30 mins.
```
Comment (mfenner1) [inline papers/yaksh/yaksh.rst]: minutes.

**Comment by mfenner1** [comment]
Summary:  I think the paper is acceptable as-is.

I'll make a few high level comments here ... I placed typo/grammatical comments directly in the PR.  Some of those comments are stylistic, so use them as you see fit.

This is a great project and should facilitate Python programming classes -- particularly large classes.  It great to see the project growing from its initial phase and to see that it is being used on increasingly large scales.  Is it possible -- or do you have a plan -- to integrate Yaksh with other open-source CMSs like Moodle (Moodle is widely used in the US -- integration with it would be helpful for, at least, US based instructors)?  Apologies if I missed that in the paper.

Three high level comments on the writing:
  
  1.  Try to simplify phrases and write more directly.  There are a few repetitions of phrases that could be shortened in places.
  2. Make sure that abbreviations are consistently capitalized
  3.  in the text, I'd prefer weight or weights to weightage (I forget where I saw this)

I would consider adding the following (but I don't consider it mandatory):

(1) add a diagram of the main components of Yaksh on the server side: grader, registry, evaluators with appropriate notes about the underlying processing systems (Tornado, Django, json, etc.) to help folks that aren't familiar with web-app components
(2) it would be useful to see an example of an assertion based test case ... it doesn't need to be the full Yaksh interface around it, but just the important code snippets

**Comment by dalippa** [comment]
@msarahan @mfenner1 Is the paper by @prabhuramachandran ready? If so, please tag us with @proceedings . Thank you!

**Comment by msarahan** [comment]
Yes, @proceedings, I think it is.

**Comment by prabhuramachandran** [comment]
@msarahan, @mfenner1, @jwjohnson314 -- thanks for the kind comments and the suggestions.  We will try to make all the suggested corrections and fixes by the 30th or so.  I hope that works.

**Comment by mfenner1** [comment]
Yes.  @proceedings.  Good to go.

**Comment by maheshgudi** [comment]
@prabhuramachandran @msarahan @mfenner1 @jwjohnson314 -- We have resolved all issues and we believe it is ready for merging. Please have a look and do the needful.

**Comment by prabhuramachandran** [comment]
I've made a final round of edits and the paper now fits in 8 pages.

**Comment by dalippa** [comment]
@mfenner1 please let us know your name so that when you attend the conference you can get a reviewer sticker. Thanks!

**Comment by mfenner1** [comment]
@dalippa Hi!  Sorry for the lag time.  Mark Fenner.  Although, I won't be making it to SciPy this year :sadface:.
